File: Stroids/src/Ship.cc
/**@file Ship
 * @author Caleb Reister <calebreister@gmail.com>
 * @brief Implementation of @see Ship
 */

#include "Ship.hh"
using namespace std;

Ship::Ship()
{
    radius = 5;
    location = WIN_MID;
    setVelocity(0, 0, 0);
    setAngle(-45);
    state = GOOD;
}

////////////////////////////////////////////////////////////////////////////
/**@fn void Ship::draw()
 * @brief Contains drawing instructions for the ship based on its current state
 * @param explodeAlpha Opacity reduction during ship explosion
 */
void Ship::draw()
{
    JSON::Array color = cfg::getInst()->read["SHIP"]["COLOR"];
    boundFix();

    //explosion variables
    const int STEPS = 100; //resolution (FPS) of explosion, dictates speed
    static int count = 1; //counts times function has been executed
    static int alpha = 255; //sets alpha (aka opacity)
    static float size = 1;

    //map the points
    ship.setPointCount(3);
    ship.setPoint(0, sf::Vector2f(10, 0));
    ship.setPoint(1, sf::Vector2f(0, 25));
    ship.setPoint(2, sf::Vector2f(20, 25));

    //check the state of the ship
    if (state == GOOD)
    {
        ship.setFillColor(sf::Color(0, 0, 0));
        ship.setOutlineColor(sf::Color(color[0].as_int(),
                                       color[1].as_int(),
                                       color[2].as_int()));
    }
    else if (state == EXPLODE)
    {
        //increment explosion variables
        count += 1;
        size += .005;
        alpha -= (255 / STEPS);

        if (count <= STEPS)
        {
            ship.setScale(size, size);
            ship.setFillColor(sf::Color(0, 0, 0, alpha));
            ship.setOutlineColor(sf::Color(color[0].as_int(),
                                           color[1].as_int(),
                                           color[2].as_int(),
                                           alpha));
        }
        else
            state = GONE;
    }
    else
    {
        ship.setFillColor(sf::Color(0, 0, 0, 0));
        ship.setOutlineColor(sf::Color(0, 0, 0, 0));
        count = 0;
    }

    ship.setOutlineThickness(cfg::getInst()->read["SHIP"]["OUTLINE_PX"].as_int());
}

////////////////////////////////////////////////////////////////////////
/**@fn void Ship::render(sf::RenderWindow& win)
 *
 * Renders the ship in the given window and gets position information from the class.
 * Does NOT get input for positioning
 *
 * @param[out] win The window in which to draw the ship
 */
void Ship::render(sf::RenderWindow& win)
{
    Vect2d loc = getLocation();

    sf::Vector2f midpoint(10, 15);
    ship.setOrigin(midpoint);
    ship.setPosition(loc.x, loc.y);
    ship.setRotation(getAngle() + 90);

    draw();
    win.draw(ship);
    updateLocation();
}

//////////////////////////////////////////////////////////////
/**@fn void Ship::applyThrust(float thrust)
 *
 * Simulates firing the engine. Changes the current velocity based on the
 * angle the ship is facing (it may be facing a different direction than it’s
 * traveling).
 *
 * @param thrust The amount of thrust to apply
 */
void Ship::applyThrust(float thrust)
{
    float rad = util::deg2rad(getAngle());
    chgVelocity(thrust * cos(rad), thrust * sin(rad),
                cfg::getInst()->read["SHIP"]["SPEED"].as_float());
}

/**@fn void Ship::explode()
 * @brief Change the state of the ship to explode, set velocity to 0.
 */
void Ship::explode()
{
    if (state != GONE)
        state = EXPLODE;
    setVelocity(0, 0, 0);
}

/**@fn ShipState getState()
 * @return The state of the ship: GOOD, EXPLODING, or GONE.
 */
ShipState Ship::getState()
{
    return state;
}

File: StroidsPy/src/Ship.py
##@file Ship.hh
#@author Caleb Reister <calebreister@gmail.com>
#@brief defines a class for controlling a space ship from above.

import math

import SpaceObject
import sfml as sf

class Ship(SpaceObject.SpaceObject):
    __ship = sf.ConvexShape()
    ##@var _state
    #This variable defines the state of the ship.
    #It will either be 'GOOD', 'EXPLODE', or 'GONE'.
    #Any other values will raise a ValueError
    #This variable should only be read by the user.
    _state = 'GOOD'
    radius = 5
    
    #################################################
    ##@fn draw()
    #@brief Contains drawing instructions for the ship based on its current state
    #@param explodeAlpha Opacity reduction during ship explosion
    def __draw(self):
        self._boundFix()
        #explosion variables
        STEPS = 100
        count = 1
        alpha = 255
        size = 1
        
        #map the points
        self.__ship.set_point_count(3)
        self.__ship.set_point(0, sf.Vector2(10, 0))
        self.__ship.set_point(1, sf.Vector2(0, 25))
        self.__ship.set_point(2, sf.Vector2(20, 25))
        #check the state
        
        if self._state is 'GOOD':
            self.__ship.set_fill_color(sf.Color(0, 0, 0))
            self.__ship.set_outline_color(sf.Color(255, 255, 255))
        elif self._state is 'EXPLODE':
            count += 1
            size += .005
            alpha -= 255 / STEPS
            if count <= STEPS:
                self.__ship.set_scale(size, size)
                self.__ship.set_fill_color(sf.Color(0, 0, 0, alpha))
                self.__ship.set_outline_color(sf.Color(255, 255, 255, alpha))
            else: self._state = 'GONE'
        elif self._state is 'GONE':
            self.__ship.set_fill_color(sf.Color(0, 0, 0, 0))
            self.__ship.set_outline_color(sf.Color(0, 0, 0, 0))
            count = 0
        else:
            raise ValueError(self._state, "is not a valid Ship state."
                                    "It needs to be either GOOD,"
                                    "EXPLODE, or GONE")
        
        self.__ship.set_outline_thickness(2)
        
    ############################################
    ##@fn render(win)
    #Renders the ship in the given window and gets position information from the class.
    #Does NOT get input for positioning
    #@param[out] win The window in which to draw the ship
    def render(self, win):
        self.__ship.set_origin(sf.Vector2(10, 15))
        self.__ship.set_rotation(self.angleDeg + 90)
        self.__ship.set_position(self.loc[0], self.loc[1])
        self.__draw()
        self.updateLocation()
        
    ##############################################
    ##@fn applyThrust(thrust)
    #
    #Simulates firing the engine. Changes the current velocity based on the
    #angle the ship is facing (it may be facing a different direction than it’s
    #traveling).
    #
    #@param thrust The amount of thrust to apply
    def applyThrust(self, thrust):
        rad = math.degrees(self.angleDeg)
        self.chgVelocity(thrust * math.cos(rad),
                         thrust * math.sin(rad),
                         max)
        
    #############################################
    ##@fn void Ship::explode()
    #@brief Change the state of the ship to explode, set velocity to 0.
    def explode(self):
        if self._state is not 'GONE':
            self._state = 'EXPLODE'
        self.setVelocity(0, 0, 0)
        
